pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -0,0 +1,253 @@
+package angulardetectorsprovider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader/angular/angulardetector"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/angularpatternsstore"
+)
+
+// backgroundJobInterval is the interval that passes between background job runs.
+// It can be overwritten in tests.
+var backgroundJobInterval = time.Hour * 1
+
+// Dynamic is an angulardetector.DetectorsProvider that calls GCOM to get Angular detection patterns,
+// converts them to detectors and caches them for all future calls.
+// It also provides a background service that will periodically refresh the patterns from GCOM.
+// If the feature flag FlagPluginsDynamicAngularDetectionPatterns is disabled, the background service is disabled.
+type Dynamic struct {
+	log      log.Logger
+	features featuremgmt.FeatureToggles
+
+	httpClient http.Client
+	baseURL    string
+
+	// store is the underlying angular patterns store used as a cache.
+	store angularpatternsstore.Service
+
+	// detectors contains the cached angular detectors, which are created from the remote angular patterns.
+	// mux should be acquired before reading from/writing to this field.
+	detectors []angulardetector.AngularDetector
+
+	// mux is the mutex used to read/write the cached detectors in a concurrency-safe way.
+	mux sync.RWMutex
+}
+
+func ProvideDynamic(cfg *config.Cfg, store angularpatternsstore.Service, features featuremgmt.FeatureToggles) (*Dynamic, error) {
+	d := &Dynamic{
+		log:        log.New("plugin.angulardetectorsprovider.dynamic"),
+		features:   features,
+		store:      store,
+		httpClient: makeHttpClient(),
+		baseURL:    cfg.GrafanaComURL,
+	}
+	// Perform the initial restore from db
+	d.log.Debug("Restoring cache")
+	if err := d.setDetectorsFromCache(context.Background()); err != nil {
+		d.log.Warn("Cache restore failed", "error", err)
+	}
+	return d, nil
```

Comment:
if you add a log line here (in case of success) you can easily get the time spent here when running this on cloud

Reply:
Good idea! Will add :+1:

Reply:
the log line already includes the timestamp so there is no need to add the duration :shrug:

Reply:
I have seen it logged like this in a few places for example:

- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/updatechecker/plugins.go#L111

- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/infra/usagestats/service/usage_stats.go#L124

- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/cleanup/cleanup.go#L120
- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/ngalert/backtesting/engine.go#L118

I guess it's good for consistency?

Reply:
yep, no problem